pkg/controller/beat/common/config.go

```diff
@@ -89,34 +96,95 @@ func buildBeatConfig(
+	userConfig, err := getUserConfig(params)
+	if err != nil {
+		return nil, err
+	}
+
 	if userConfig == nil {
 		return cfg.Render()
 	}
 
-	userCfg, err := settings.NewCanonicalConfigFrom(userConfig.Data)
+	if err = cfg.MergeWith(userConfig); err != nil {
+		return nil, err
+	}
+
+	return cfg.Render()
+}
+
+// getUserConfig extracts the config either from the spec `config` field or from the Secret referenced by spec
+// `configRef` field.
+func getUserConfig(params DriverParams) (*settings.CanonicalConfig, error) {
+	if params.Beat.Spec.Config != nil {
+		userCfg, err := settings.NewCanonicalConfigFrom(params.Beat.Spec.Config.Data)
+		if err != nil {
+			return nil, err
+		}
+		return userCfg, nil
+	}
+
+	if params.Beat.Spec.ConfigRef == nil {
+		return nil, nil
+	}
+
+	secret, err := getConfigRefSecret(params.Client, params.Beat.Spec.ConfigRef.SecretName, params.Beat.Namespace)
 	if err != nil {
 		return nil, err
 	}
 
-	if err = cfg.MergeWith(userCfg); err != nil {
+	nsn := k8s.ExtractNamespacedName(&params.Beat)
+	if err := watches.WatchUserProvidedSecrets(nsn, params.DynamicWatches(), ConfigRefWatchName(nsn), []string{secret.Name}); err != nil {
 		return nil, err
 	}
-	log.V(1).Info(
-		"Replacing ECK-managed configuration by user-provided configuration",
-		"beat_name", beat.Name,
-		"namespace", beat.Namespace,
-	)
 
-	return cfg.Render()
+	data, exists := secret.Data[ConfigFileName]
+	if !exists {
+		msg := fmt.Sprintf("no %s key in secret", ConfigFileName)
+		params.Logger.Error(err, msg, "namespace", params.Beat.Namespace, "beat_name", params.Beat.Name, "secret_name", secret.Name)
+
+		msg = fmt.Sprintf("no key %s in secret %s in namespace %s", ConfigFileName, secret.Name, params.Beat.Namespace)
+		params.Recorder().Event(&params.Beat, corev1.EventTypeWarning, events.EventReasonUnexpected, msg)
+		return nil, fmt.Errorf(msg)
+	}
+	parsed, err := settings.ParseConfig(data)
+	if err != nil {
+		msg := "unable to parse configuration from secret"
+		params.Logger.Error(err, msg, "namespace", params.Beat.Namespace, "beat_name", params.Beat.Name, "secret_name", secret.Name)
+
+		msg = fmt.Sprintf("unable to parse configuration from key %s in secret %s in namespace %s", ConfigFileName, secret.Name, params.Beat.Namespace)
+		params.Recorder().Event(&params.Beat, corev1.EventTypeWarning, events.EventReasonUnexpected, msg)
+		return nil, err
+	}
+	return parsed, nil
+}
+
+func getConfigRefSecret(
+	client k8s.Client,
+	name, namespace string,
+) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	if err := client.Get(
+		types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		}, secret); err == nil {
+		// user provided an existing secret, just use it
+		return secret, nil
+	} else if !apierrors.IsNotFound(err) {
+		// can't be sure if the secret exists or not, error out to be safe
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("secret %s not found in %s namespace", name, namespace)
```

Comment:
I would keep the original error, it would keep the code simpler and the original message is quite explicit: `Secret "foo" not found`